gtsam/slam/EssentialMatrixFactor.h

```diff
@@ -267,28 +269,121 @@ class EssentialMatrixFactor3: public EssentialMatrixFactor2 {
       EssentialMatrix cameraE = cRb_ * E;
       // Evaluate error
       return Base::evaluateError(cameraE, d, boost::none, Dd);
     } else {
       // Version with derivatives
-      Matrix D_e_cameraE, D_cameraE_E; // 2*5, 5*5
+      Matrix D_e_cameraE, D_cameraE_E;  // 2*5, 5*5
       EssentialMatrix cameraE = E.rotate(cRb_, D_cameraE_E);
       Vector e = Base::evaluateError(cameraE, d, D_e_cameraE, Dd);
-      *DE = D_e_cameraE * D_cameraE_E; // (2*5) * (5*5)
+      *DE = D_e_cameraE * D_cameraE_E;  // (2*5) * (5*5)
       return e;
     }
   }
 
-public:
+ public:
   GTSAM_MAKE_ALIGNED_OPERATOR_NEW
 };
 // EssentialMatrixFactor3
 
-}// gtsam
+/**
+ * Factor that evaluates algebraic epipolar error (K^-1 p)'E (K^-1 p) for given
+ * essential matrix and calibration. The calibration is shared between two
+ * images.
+ */
+template <class CALIBRATION>
+class EssentialMatrixFactor4
+    : public NoiseModelFactor2<EssentialMatrix, CALIBRATION> {
+ private:
+  Point2 pA_, pB_;  ///< points in pixel coordinates
+
+  typedef NoiseModelFactor2<EssentialMatrix, CALIBRATION> Base;
+  typedef EssentialMatrixFactor4 This;
+
+  static const int DimK = FixedDimension<CALIBRATION>::value;
+  typedef Eigen::Matrix<double, 2, DimK> JacobianCalibration;
+
+ public:
+  /**
+   *  Constructor
+   *  @param keyE Essential Matrix variable key
+   *  @param keyK Calibration variable key
+   *  @param pA point in first camera, in pixel coordinates
+   *  @param pB point in second camera, in pixel coordinates
+   *  @param model noise model is about dot product in ideal, homogeneous
+   * coordinates
+   */
+  EssentialMatrixFactor4(Key keyE, Key keyK,
+                         const Point2& pA, const Point2& pB,
+                         const SharedNoiseModel& model)
+      : Base(model, keyE, keyK), pA_(pA), pB_(pB) {}
+
+  /// @return a deep copy of this factor
+  gtsam::NonlinearFactor::shared_ptr clone() const override {
+    return boost::static_pointer_cast<gtsam::NonlinearFactor>(
+        gtsam::NonlinearFactor::shared_ptr(new This(*this)));
+  }
+
+  /// print
+  void print(
+      const std::string& s = "",
+      const KeyFormatter& keyFormatter = DefaultKeyFormatter) const override {
+    Base::print(s);
+    std::cout << "  EssentialMatrixFactor4 with measurements\n  ("
+              << pA_.transpose() << ")' and (" << pB_.transpose() << ")'"
+              << std::endl;
+  }
+
+  /**
+   * @brief Calculate the algebraic epipolar error p' (K^-1)' E K p.
+   *
+   * @param E essential matrix for key keyE
+   * @param K calibration (common for both images) for key keyK
+   * @param H1 optional jacobian of error w.r.t E
+   * @param H2 optional jacobian of error w.r.t K
+   * @return * Vector 1D vector of algebraic error
+   */
+  Vector evaluateError(
+      const EssentialMatrix& E, const CALIBRATION& K,
+      boost::optional<Matrix&> H1 = boost::none,
+      boost::optional<Matrix&> H2 = boost::none) const override {
+    Vector error(1);
```

Comment:
Declare where used, not before. And make const.

Reply:
```const Vector error(1);
error << E.error(vA, vB, H1) 
```
does not compile. 

```
 candidate: ‘Eigen::CommaInitializer<Derived> Eigen::DenseBase<Derived>::operator<<(const Scalar&) [with Derived = Eigen::Matrix<double, -1, 1>; Eigen::DenseBase<Derived>::Scalar = double]’ <near match>
  144 | inline CommaInitializer<Derived> DenseBase<Derived>::operator<< (const Scalar& s)
      |                                  ^~~~~~~~~~~~~~~~~~
passing ‘const Vector*’ {aka ‘const Eigen::Matrix<double, -1, 1>*’} as ‘this’ argument discards qualifiers

```